[PR-1190] Article about typescript for haskellers

sashasashasasha151 opened the pull request.

sashasashasasha151 committed:
[PR-1190] Article about typescript for haskellers

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+# TypeScript for Haskellers
+
+Web app nowadays is mandatory part of every modern application.
```
This is a bit too much water: the question here is why Haskellers would be interested in TypeScript.

NaeosPsy replied:
Additionally: could you drop a link to our Typescript article somewhere in the body (https://serokell.io/blog/why-typescript). This is for SEO reasons, but I am quite sure a good spot can be found.

sashasashasasha151 replied:
>This is a bit too much water

Yes, started it as I usually do my university docs) Will fix

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+As in the previous parts you can also see more complex examples in Yuriy Bogomolov [edsl workshop](https://github.com/YBogomolov/workshop-edsl-in-typescript/tree/master) and Giulio Canti [example](https://gist.github.com/gcanti/9a0c2a666621f03b80457831ff3ab997).
+
+## Conclusion
```
Perhaps add a summary of the article here

sashasashasasha151 replied:
Hm, what kind of summary it should be?

NaeosPsy replied:
Rather generic: what you talked about in the article, some main points repeated, perhaps some links to other places.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+We also will not say about disadvantages of TypeScript type system against Haskell or how it works under the hood.
+Our goal is to show what and how you can use types in TypeScript.
+
+## Similar concepts
```
Since we have a section about similar concepts, perhaps there are some unsimilar pitfalls that Haskell developers might fall into? 

It would be a nice "mirror section" for this one.

sashasashasasha151 replied:
Hm, I think I should just somehow rename the section as all article is about `Similar concepts` but this part is just about easiest things

> It would be a nice "mirror section" for this one

I didn't want to mention them because in this case it will become a TypeScript guide but not a brief introduction to types in it. I also didn't want to describe all code and how it work, only examples, so that reader may go and google it and find more examples.

Martoon-00 commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+Unions are not tagged, so it is just a condition of possible types.
+
+```typescript
+type Result = string | number | (() => string);
````
At this point I wonder, whether can I mix unit and union types, e.g.

```ts
type Ty = string | false
```

and if no, why can it be made this way (if there is a short explanation).

sashasashasasha151 replied:
Yes, you can

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+    case "result":
+      return `valid ${result.value.toString()}`;
+    default:
+      return "impossible";
```
It looks suspicious to me that the compiler does not consider `default` case as impossible :thinking:

sashasashasasha151 replied:
Hm, I think the answer here is that TypeScript has type `never` and the last case here is with the type `never`. There is also a rule `default-case` about always enable default case and it case of enabling error on impossible `default` case they will conflict. So, if you will read some issues answers in their github you will see that they usually don't like popular ideas

Martoon-00 replied:
Hah, nice

Martoon-00 commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+    return undefined;
+  }
+}
+```
````
IMO It is important to demonstrate somewhere soon, that TS compiler can into exhaustiveness check, that for `if (typeof arg === "string")` in `else` branch the compiler already knows that `arg` cannot be `string`. Overwise the whole story about TS's type-system stops making much sense.

Also, where `arg` and `n` variables came from? :eyes:

Martoon-00 commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+  .with({ type: 'ok', data: { type: 'text' } }, (res) => `<p>${res.data.content}</p>`)
+  .with({ type: 'ok', data: { type: 'img', src: select() } }, (src) => `<img src=${src} />`)
+  .exhaustive();
+```
````
:pog: Looks interesting

Heimdell replied:
Looks like something that eats stack.

Martoon-00 commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+a["y"] = 1; // error
+```
+
+Also be careful passing object with readonly fields to function as it may cause unexpected behaviour.
````
I would say it more clearly, that the passed object can be changed inside (I personally would expect the call to fail, hilarious).

If it is possible to make the text simultaneously short and self-contained (leaving code snippets as an optional extra to read), I think this should be done.

sashasashasasha151 replied:
As I understand that is because of scope and that is a problem. So, function know nothing about `arg` type beyond itself. It know only that `arg: { x: number }`. But maybe not, will try to understand it more

Heimdell replied:
If `readonly` is a part of the type, `{readonly x : A}` should be a subtype of `{x : A}` and therefore latter should be prohibited when expecting former.

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+const nextA = produce(a, (draft) => {
+  draft[1].valid = false;
+  draft.push({ text: "some text", valid: false });
```
Spent 30secs figuring out, what can `push` mean as a lens-like operation, before remembering that it is an Array built-in function. Would like to see a comment here, many Haskellers may not know JavaScript at all :stuck_out_tongue_closed_eyes:

Heimdell replied:
How can you _`push()`_ and _assign_ into an immutable `A = Immutable<_>`? Ah, so the `produce` lowers the immutability for a callback? Plz add that as a comment.

Martoon-00 replied:
AFAIG `product` creates a partial clone with the following changes applied, the old value is left intact. Is that right?

sashasashasasha151 replied:
>AFAIG product creates a partial clone with the following changes applied, the old value is left intact. Is that right?

During my research I saw great explanation about `immer` implementation under the hood. But now I can't find it. So, yes, I think on high level this answer is correct (it is what we can know from their docs)

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+const arrLength: <T>(_: T[]) => number = (arr) => arr.length;
+
+arrLength<number>([1, 2, 3]);
+arrLength<string>(["1", "2", "3"]);
```
It becomes harder to understand what's going on. E.g. what `arrLength` returns (type or value?), where can I use it?

Heimdell replied:
Is it necessary to write the type each time you use the function? If not, please add an example.

sashasashasasha151 replied:
>It becomes harder to understand what's going on. E.g. what arrLength returns (type or value?), where can I use it?

Hm, in the previous part about `currying` there was an example `add: addT`. It show how we can split type and implementation and also the syntax. \
So this is like 
```haskell
arrLength :: [a] -> Int
arrLength l = length l
```
But written smth like `arrLength :: [a] -> Int = l -> length l` (It is funny that students on Haskell course were asking about such syntax) \

> where can I use it

Everywhere? It is just a function (maybe not really useful)

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  };
+};
+
+f("5")(5);
```
Here and further, could you also mention (e.g. in comments) what is returned/outputed by the call?

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+    if (cmp.equal(k, kk)) {
+      result = v;
+    }
+  });
```
In this particular case, is that true that you could go without the `result` variable and just do `return v` when the value is found?

sashasashasasha151 replied:
`return` in `foreach` make nothing, just didn't want to use `for`, but I will try to make a better example

Heimdell replied:
I'd make it via `Array#reduce`. @Martoon-00, the `return` will just exit the `forEach`'s lambda.

Martoon-00 replied:
Ah right, fair enough.

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  equal(f: number[], s: number[]) {
+    return f.filter(item => s.indexOf(item) < 0).length === 0;
+  }
+}
```
In real life, it is a bit dangerous to define `equal` that is not symmetric, but if you like this example then fine.

I would only reflect this behavriour in the class name somehow, maybe call it `IntArrSubset`?

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+let a: A = {};
+const b: B = {x: 1};
+a = b;
+console.log(a.x); // Property 'x' does not exist on type 'A'.
```
Is `Property 'x' does not exist on type 'A'.` the outputted text or compile-time error? If it is the latter, I would simplify to `const c = a.x` to make it clearer.

Heimdell replied:
:cho:

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+It allows you to abstract by the record fields and allow only those of them which have such fields.
+Unfortunately TypeScript type compatibility is based on structural subtyping, and it is not a row polymorphism.
+Despite this we can use intersection types to emulate this.
+But you should be careful with using it because subtyping cause information loss as if you have value `a: T` and `b: T & {x: number}` you will be able to do `a = b` and cut off `b` to `T`.
```
The example is too concrete and complex IMO, hard to grasp. I would just say that additional information can be lost during assignments, the developer must be cautious to propagate the full type-level information when they need it.

Heimdell replied:
Wait, the `a = b` is a reference assignment, the `b` will still be `T & {x: number}` internally, it will only appear `T`, won't it?

sashasashasasha151 replied:
> Wait, the a = b is a reference assignment, the b will still be T & {x: number} internally, it will only appear T, won't it?

So, yes, but compiler will not allow you to call `x`

Martoon-00 commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+console.log(a.x); // Property 'x' does not exist on type 'A'.
+```
+
+### Mapped types, Conditional types and Type families
````
Here some advanced stuff comes in, maybe leave it for the second part of the article? I believe, at this moment the reader must have a plenty of thoughts to digest, so it could make sense to wrap up and say that more interesting things like mapped types, peano numbers and others will be considered in the next article.

@NaeosPsy What do you think?

NaeosPsy replied:
I think that it can definitely be in one article, as long as easier things are first -- the person will read the start, hit a wall, and bookmark the article for re-read later.

Martoon-00 replied:
Okay, I'll review the remaining part a bit later then.

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+Web app nowadays is mandatory part of every modern application.
+It doesn't matter how small or complex it is.
+From one click apps with pictures converting to photoshop everyone wants fast and easy access to it and Web is where you can give it to your users or even for you with the easiest way.
+
```
In some projects, we had a practice that no documentation should be published unless verified with some spellchecker. I also tried to follow it for blog posts, and believe it saved a considerable amount of time.

So I'd really recommend registering in e.g. https://grammarly.com and making the pass over the full text there.

Martoon-00 reviewed (commented):
I reviewed about 70% of the text, so far so good, really interesting info.

Left some comments for now.

Also, I think at some point later we will need more reviewers on the technical part because I find some points in the storytelling/examples/explanations controversial (aside from those that are mentioned in the review).

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+It doesn't matter how small or complex it is.
+From one click apps with pictures converting to photoshop everyone wants fast and easy access to it and Web is where you can give it to your users or even for you with the easiest way.
+
+Moreover, UI and UX standards are increasing.
```
_Raising_ or _improving_

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+From one click apps with pictures converting to photoshop everyone wants fast and easy access to it and Web is where you can give it to your users or even for you with the easiest way.
+
+Moreover, UI and UX standards are increasing.
+People like simple interfaces with enjoyable.
```
with joy

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+Application with worse functionality and better UI/UX may win on the market against the other one with bad design.
+
+One might say that doesn't matter what your application is about or what part of it you write.
+In the end in most of the cases you will need to give access for it to smb.
```
```suggestion
In the end you'll have to give access for it to somebody other.
```

Heimdell replied:
I suggest replacing dot with comma and gluing the next sentence.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+In the end in most of the cases you will need to give access for it to smb.
+Whether it is an external user or your colleague using internal products.
+
+So, in Serokell we use TypeScript for writing web pages.
```
> So

This implies that you explained why the TypeScript is an only choice, which haven't really happened.

```suggestion
In Serokell, we use TypeScript for writing web pages.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+Moreover, these 3 things are just files, and you need a way to read and show them.
+That is what browsers do.
+
+Historically this set of files was perfect to write web pages.
```
```suggestion
Historically, this set of files was perfect to write web pages.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+That is what browsers do.
+
+Historically this set of files was perfect to write web pages.
+And because of the enormous growth of the web JavaScript in particular has grown enormous as well.
```
```suggestion
And because of the enormous growth of the web, JavaScript in particular has grown enormous as well.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+And there are a lot of such languages and language extensions: TypeScript, Kotlin JS, CoffeeScript, Scala JS, Babel, ClojureScript, etc.
+Generally you can write web applications almost on every language.
+
+As we like Haskell, maybe we can write our Web application on functional languages or even on Haskell? And the answer is yes.
```
```suggestion
Because we like Haskell, maybe we can write our Web application on functional languages or even on Haskell? And the answer is yes.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+But when we use Haskell we know that people like it and customers want it, and we can't say the same thing about technologies above.
+Customers will be locked in this solution as most likely they will not be able to maintenance such application by themselves.
+On our side we will be locked in using JS libraries which number is enormous, build problems, and fast team expanding.
```
> JS libraries which number is enormous

Wut? I thought the current case is the "GHCJS" one, because

> Customers will be locked in this solution as most likely they will not be able to maintenance such application by themselves.

sashasashasasha151 replied:
Okay, maybe I am not right, but for example. Smb wrote JS library yesterday. Am I able to use it with GHCJS?

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+There is strongly-typed PureScript with Haskell like syntax, purely functional Elm, Reflex framework for Haskell with GHCJS, F# with Fabel.
+
+But when we use Haskell we know that people like it and customers want it, and we can't say the same thing about technologies above.
+Customers will be locked in this solution as most likely they will not be able to maintenance such application by themselves.
```
```suggestion
Customers will be locked in this solution because most likely they will not be able to maintain such an application by themselves.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+Thus, we needed a language which will easy support all modern technologies from web development, big community, and give us a possibility to use our knowledge from Haskell.
+And here it is -- TypeScript.
+It is static typing language which popularity grows from year to year.
+A lot of new libraries are writen on it, a lot of have type declarations, and it provides an easy way to use plain JS libraries as well.
```
```suggestion
A lot of new libraries are writen on it or have type declarations, and TypeScript provides an easy way to use plain JS libraries as well.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+## Similar concepts
+
+Unlike Haskell in TypeScript types are not the center part of the language.
+They just help programmers to write more strict and understandable code, making their life easier.
```
```suggestion
They just help programmers to write stricter and more understandable code, making their life easier.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+They just help programmers to write more strict and understandable code, making their life easier.
+You can use types like `any` and nobody pushes you into strict frames.
+
+So, this part will be about simple build in concepts which are familiar for haskellers and which do not require any overheads on writing code.
```
```suggestion
So, this part will be about simple built-in concepts which are familiar for haskellers and do not require any overheads on writing code.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+You can use types like `any` and nobody pushes you into strict frames.
+
+So, this part will be about simple build in concepts which are familiar for haskellers and which do not require any overheads on writing code.
+Nevertheless, we will also leave links for some more complex libraries and articles which may help you or deeply investigate the theme.
```
```suggestion
Nevertheless, we will also leave links for some more complex libraries and articles which may help you investigate the case even deeper.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+### Strictness
+
+The first thing you need to know is that TypeScript was developed to provide an ability for JavaScript developers using it.
```
Не понимат

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+The first thing you need to know is that TypeScript was developed to provide an ability for JavaScript developers using it.
+So, by default TypeScript is really polite in using types.
+Nevertheless, with bunch or compiler options you can make it more strict in types and other things.
+So, the main strict compiler option is which you need to add to your `tsconfig.json` file is `strict`.
```
```suggestion
So, the main compiler option is which you need to add to your `tsconfig.json` file is `strict`.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+So, by default TypeScript is really polite in using types.
+Nevertheless, with bunch or compiler options you can make it more strict in types and other things.
+So, the main strict compiler option is which you need to add to your `tsconfig.json` file is `strict`.
+It will enable all strict related flags for you and if in future smth will be added to strict options list you will not need to add it manually.
```
```suggestion
It will enable all strictness-related flags for you and if in future something will be added to strict options list you will not need to add it manually.
```
Smone usd smth in artcl, plz dn't.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+Nevertheless, with bunch or compiler options you can make it more strict in types and other things.
+So, the main strict compiler option is which you need to add to your `tsconfig.json` file is `strict`.
+It will enable all strict related flags for you and if in future smth will be added to strict options list you will not need to add it manually.
+But if you want to specify them here are all options: alwaysStrict, strictNullChecks, strictBindCallApply, strictFunctionTypes, strictPropertyInitialization, noImplicitAny, noImplicitThis, useUnknownInCatchVariables.
```
```suggestion
But in case you want to specify them all, here you are: alwaysStrict, strictNullChecks, strictBindCallApply, strictFunctionTypes, strictPropertyInitialization, noImplicitAny, noImplicitThis, useUnknownInCatchVariables.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+### Type aliases and newtypes
+
+Type aliases works exactly the same as in Haskell.
```
```suggestion
Type aliases works exactly like they do in Haskell -
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+### Type aliases and newtypes
+
+Type aliases works exactly the same as in Haskell.
+You can create a new name to refer to some type.
```
```suggestion
you can create a new name to refer to some type.
```

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+}
+```
+
+Creating newtype is not so trivial.
````
```suggestion
Creating a `newtype` is not so trivial.
```

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+sendMessage("email@gmail.com" as Email);
+```
+
+Anyway this is not an ideal solution for newtypes but there is a [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in issue with a better solution and there is also a [library](https://github.com/sveseme/taghiro) which implement it to a nice view.
````
```suggestion
Anyway, this is not an ideal solution for newtypes but there is a [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in issue with a better solution and there is also a [library](https://github.com/sveseme/taghiro) which implement it to a nice view.
```

Heimdell replied:
> implement it to a nice view.

:cho:

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+### Unit types, Unions, Data types and Pattern Matching
+
+Haskell ADT is a constantly used functionality of language.
```
s/./, and/

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+}
+```
+
+Going deeper using Unions it is possible to construct data type analog.
````
```suggestion
Going deeper using Unions, it is possible to construct data type analog.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+}
+
+changeA(a);
+console.log(a.x); // 2
```
Noice.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  return newArray; // or just simple return arr.filter(fn);
+};
+
+const f = (x: string): ((s: number) => boolean) => {
```
What happens if a parameter name in type and in actual returned function diverge?

sashasashasasha151 replied:
Nothing) \
Parameter names in types are only needed for, hm, nothing and it is just impossible to not write them \
You can define your function like 
```typescript
const f: (_: number) => string = (x) => {
  return x.toString();
};

const g = (x: number): string => {
  return x.toString();
};

function h(x: number): string {
  return x.toString();
}
```
In the first case you just write the type and then implementation. In the second and third ones parameters and their types are combined. In the example you mentioned the return type is just type and we can simply write `const f = (x: string): ((_: number) => boolean) => {`

sashasashasasha151 replied:
If you want smth like real names you need to write smth like
```typescript
const f = (x: string): ((_: { s: number }) => boolean) => {
  return ({ s }: { s: number }) => {
    return s.toString() === x;
  };
};
```

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+So, yes, we can implement some king of currying using it.
+
+```typescript
+type addT = (_: number) => (_: number) => number;
````
Can we do a parametrised type alias?

like
```ts
type curried<A, B, C> = (_: A) => (_: B) => C
```

sashasashasasha151 replied:
Yes, there are examples below

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+### Polymorphism
+
+There are different polymorphisms, but we will say only about thee of them: Parametric, Ad hoc, and Row.
```
```suggestion
There are different polymorphisms, but we will say only about thee of them: Parametric, Ad-hoc one and Row one.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+type addT = (_: number) => (_: number) => number;
+const add: addT = (l) => (r) => l + r;
+
+type foldArrT = (f: (_: number) => (_: number) => number) => (_: number) => (_: number[]) => number;
```
This thing also screams "add type parameters to me", because it will work for any array, not just for `number[]`.

sashasashasasha151 replied:
Yes, but only on the next step we say about `Polymorphism`. I can just swap this two parts, but not sure that it will be better

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+### Polymorphism
+
+There are different polymorphisms, but we will say only about thee of them: Parametric, Ad hoc, and Row.
+Let's start from the first on.
```
```suggestion
Let's start from the first one.
```

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+lookup(new IntArrEq(), [1, 2, 3], [[[1, 3], "1"], [[3, 2], "2"], [[2, 1, 3], "3"]]);
+```
+
+One more polymorphism is row.
````
```suggestion
Another kind of polymorphism is row one.
```

Heimdell replied:
I think all three kinds of polymorphism should get their subheaders.

sashasashasasha151 replied:
You mean in `Polymorphism` header or remove `Polymorphism` and add three new headers?

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+```
+
+One more polymorphism is row.
+It allows you to abstract by the record fields and allow only those of them which have such fields.
````
```suggestion
It allows you to make sure that the record contains at least given set of fields.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+One more polymorphism is row.
+It allows you to abstract by the record fields and allow only those of them which have such fields.
+Unfortunately TypeScript type compatibility is based on structural subtyping, and it is not a row polymorphism.
```
```suggestion
Unfortunately TypeScript type compatibility is based on structural subtyping, so what we have available it is not _really_ a row polymorphism.
```

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+
+```typescript
+type fnT = <T>(v: T & { x: number }) => T & { x: number }
+// You should fully describe all types to not lose information.
````
In Elm for instance, you can (or was able to) write a `{x : Int | t} -> t` function and it will not compose with itself as well. I would not say that this is information loss, but rather a chosen level of abstraction. Sometimes you _need_ to forget something - or outright remove a field.

sashasashasasha151 replied:
I think in `Elm` you just can't return `x` from the function with type `{x : Int | t} -> t` \
I mean you can't write such thing
```elm
type alias A =
  { x : Int
  , y : Int
  }

type alias B t = {t | x : Int}
  
f : B t -> t
f a = a
```

But in TS it is possible
```typescript
type A = { x: number; y: number };

type B<T> = T & { x: number };

const f: <T>(a: B<T>) => T = (a) => a;
```
So, in case of `Elm` yes, we want to remove field `x` as if we will not remove it compiler will say about it to us. But in case of TS it will say nothing.

sashasashasasha151 replied:
>remove a field

Moreover it will not remove it at all but only on type level.
```
type A = {};
type B = A & { x: number };

let a: A = {};
const b: B = { x: 1 };
a = b;
console.log(JSON.stringify(a)); // {"x":1}
```

sashasashasasha151 replied:
So, I think `const f: <T>(a: B<T>) => T = (a) => a;` may cause bugs and you should `fully describe all types to not lose information`

sashasashasasha151 replied:
>fully describe all types

I think it is the biggest problem with TS when you try to go a little bit deeper. Compiler will help you only if you will help him

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+Mapped types allows you to create new types based on old types transforming them in some way.
+The syntax looks like `[set of field names]: type;`.
+For example, you can add optional modifier or readonly modifier to each field with `type Partial<T> = { [P in keyof T]?: T[P] }` and `type Readonly<T> = { readonly [P in keyof T]: T[P] }`.
+It is also possible to remove such modifiers using `-?` and `-readonly`.
```
Please provide examples of that.

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+Let's present some examples of mapped types usage.
+
+```typescript
+const sym: unique symbol = Symbol();
````
Wait, what's `unique`?

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+type Point = {
+  x: number;
+  y: number;
+  [sym]: string;
```
The usage of symbols as _computed names_ and the whole _computed name_ concept should be described beforehand. When I first found I can do `{[x + "1"] : "a"}` in JS I was like _NANI?!_ for 20 minutes or so.

This is a very alien thing to have a computed field in what appears to be a record. For me, at least.

sashasashasasha151 replied:
Okay, I added symbols just to not write comments, but as I already started doing it I will simply remove them. I think computed names part is not for this article.

sashasashasasha151 replied:
> computed names part is not for this article

Okay, `[P in keyof T]?: T[P]` is already computed names 🤔 \
Do we really need a part about them?

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  0: string;
+}
+
+type Pick<T, K extends keyof T> = {
```
This thing should also have a "here we..." explanation above it. I tried to understand what is it and I got lost.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+type Getter<T extends string> = `get${Capitalize<T>}`;
+type Getters<T> = {
+    [K in keyof T as Getter<Extract<K, string>>]: () => T[K];
```
_Magic intensifies_

Heimdell replied:
What's the semantic of `in`, `keyof` and `as` in types? Please describe beforehand. Okay, for `in` (subtyping) and `keyof` (javascript) it is kinda obvious. But `as`?

Heimdell replied:
After some thought I got that it is a "mapped type", but please explain, because if I was reading this not for a review, I'd "okay, I'm out, `Ctrl-W`" at least here. Like "using `as` we can do ... and here ls an example".

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+Another great feature is conditional types.
+It allows you to choose the type based on some condition expressed as another type.
+Using `extends` word you are able to make a condition `T extends U ? X : Y`.
```
What's the priority of `extends` and `?:` operators?

sashasashasasha151 replied:
Does it really matter? (I just can't find this information 🙂)

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  z: number;
+}
+
+type PointX = Omit<Point, "y" | "z">; // type PointX = { x: number }
```
What exactly are we trying to do in this code block? I got that the `never` is a lower boundary of a type lattice :peka:, but I can't understand that the code is doing, so it need an explanation: "here we have a type `Point`, but we want `X`, so we use `Y` to ... and `Z` to ...".

NaeosPsy replied:
I think it would be simpler to write PointNotYZ or smth like that. 

Or just make a code sample with Point2D and Omit z

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+## Going deeper
+
+Build in concepts are good.
```
```suggestion
Built-in concepts are good.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+## Going deeper
+
+Build in concepts are good.
+But as you already saw TypeScript type system is not really weak.
```
```suggestion
But as you can see, TypeScript type system is not really a weak one.
```

Heimdell replied:
It is rather intimidatingly flexible.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+But as you already saw TypeScript type system is not really weak.
+Adding our knowledge from Haskell and type theory we can think about how to get more from this system and what we can build using it.
+This part will show very brief descriptions of different approaches.
+But links on references and other articles must help you to deeply understand them.
```
:cho:

Maybe I don't get it, but this sentence doesn't tell where the links _are_.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+### HKT
+
+You may see that TypeScript type system is not so bad as one may think.
+But it has one important limitation -- absence of kinds.
```
_And absence of_ soundness, _but okay._

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+You may see that TypeScript type system is not so bad as one may think.
+But it has one important limitation -- absence of kinds.
+Higher-Kinded Types allow us to write types which have their own type constructors.
```
I'd call that "higher-order functions _over types_". The term "type constructor", I believe, got a bit outdated. And `type List<A>` is a type constructor already, with no HKT.

Or "allow types to be parametrized by other parametrized types".

or
```suggestion
Higher-Kinded Types allow us to write types which have their own type constructors as parameters.
```

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+
+```typescript
+interface Collection<F> {
+  create: <A>() => F<A>; // Type 'F' is not generic.
````
Can we abstract over generics or not? The paragraph above say we can't, but the code says we can. Does this code work? If yes, what happens if we do `Collection<number>`? If not, what is the compiler error? The comment `Type 'F' is not generic.` is very misleading.

sashasashasasha151 replied:
>The comment Type 'F' is not generic. is very misleading.

It is compiler error)

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+};
+```
+
+Fortunately we are able to simulate kinds using denationalization which allows us to translate higher-order programs into a first-order language.
````
> [denationalization](https://dictionary.cambridge.org/ru/%D1%81%D0%BB%D0%BE%D0%B2%D0%B0%D1%80%D1%8C/%D0%B0%D0%BD%D0%B3%D0%BB%D0%B8%D0%B9%D1%81%D0%BA%D0%B8%D0%B9/denationalization)

Please don't invoke political operators :D

Defunctionalisation, perhaps?

sashasashasasha151 replied:
Oru, macbook touchbar just suggested this word I think)))

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  'Map': Map<A, B>;
+}
+
+type URIS = keyof URItoKind<unknown>;
```
What's an `unknown`?

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+type URIS = keyof URItoKind<unknown>;
+type URIS2 = keyof URItoKind2<unknown, unknown>;
+
+type Kind<F extends URIS, A> = URItoKind<A>[F];
```
~What is going on here? How does it work?~ I get it now, but it should be described verbatin - "we create interface A so we can later B".

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+  create: () => [],
+  insert: (v) => (c) => [...c, v]
+};
+```
````
Ok, so the previous version did not compile.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+}
+
+interface URItoKind2<A, B> {
+  'Map': Map<A, B>;
```
So, it is a _closed_ relation? I assume we can't add type constructors to the system to be recognised by the defunctionaliser.

sashasashasasha151 replied:
Hm, you mean if it is possible to create our own type and add them here? We can do it, examples are in `Another way to implement eDSL is tagless final.` part. This one is just our manually created minimal examples without libs just to show how it works. Or maybe I didn't understand you

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+type Fibonacci<N, F0 = Zero, F1 = One> = {
+  acc: F0
+  n: N extends Succ<infer _> ? Fibonacci<Decrement<N>, F1, Add<F0, F1>> : never
+}[IfElse<Equals<Zero, N>, "acc", "n">]
```
_Why._

Okay. So this thing is outright made from high-circle magic spells. What's `infer _`?

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+### GADT and eDSL
+
+Generalized algebraic data types in Haskell give us an ability to manually write types of the constructors.
```
No, they give us ability to construct/match objects of type `Foo a b c` with some of `a`, `b` or/and `c` constrained to be certain types or to have constraints like `Eq a` or `Action a c`.

For
```haskell
data Exp a where
  Var :: Name -> Exp Int
  Lam :: Name -> Exp a -> Exp (Int -> a)
  App :: Exp (a -> b) -> Exp a -> Exp b
```
You can't construct general `Exp b` - the `b` by definition must be an `Int` or a function.
When you match an `Exp` and get a `Var` you know that it's type is `Exp Int`.

Which allows you to write
```haskell
eval :: Map String Int -> Exp a -> Maybe a
eval scope (Var n) = Map.lookup n scope
eval scope ... = ...
```
In the first clause of the function we get local `a ~ Int` constraint, and therefore we _have to_ return `Maybe Int`.

and the resulting `a` will be the same as the `a` carried by the concrete `Exp` you plug into the function.

----

"Manually writing" is just a consequence.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+Generalized algebraic data types in Haskell give us an ability to manually write types of the constructors.
+Having data type `D a` with type value `a` we are able to create constructors like `C :: Int -> D Int`.
+The key feature here is that Haskell may say us about equality of types `a` and `Int` here.
```
I'm sorry for the rant above ._.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+import { identity } from "fp-ts/lib/function";
+
+interface Equality<A, B> {
+  (a: A): B;
```
Wut. Plz explain the magics.

Heimdell replied:
Does that mean, that `Equality` should be a funciton `A -> B`?

Heimdell replied:
Ah, I see

sashasashasasha151 replied:
yes

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  | { type: "And"; l: ArithExpr<boolean>; r: ArithExpr<boolean>; proof: Equality<boolean, A> };
+
+const num: (v: number) => ArithExpr<number> = (v) => {
+  return { type: "Num", v, proof: identity };
```
The `Equality`/`identity` trick should be explained.

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+We will use `fp-ts` to implement it.
+
+```typescript
+import { Kind, URIS } from "fp-ts/HKT";
````
What's the `Kind` here? How is it [un]related to haskell's definition of `Kind` (type of types)?

The `URIS` sounds like `URI`. Please expain what it is.

Martoon-00 replied:
The scope of the post is really large, the difficulty of described concepts grows exponentially throughout the text, so many such issues arise.

But need to say, things that are now explained (like `Kind`) looked explained quite well to me and embedded into the text nicely. At least those that are similar to what I have already worked with in Haskell. Other things still look mind-blowing, but that's probably exactly the scope of the post to only provide the transition from known Haskell concepts to TS.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+
+type ToS<A> = {
+  toString: string;
+};
```
Shouldn't `ToS` be `Interpreter<String>`?

sashasashasasha151 replied:
Hm, the idea was to take example from our fp course https://slides.com/fp-ctd/lecture-11#/18/0/1

And as I understand the problem (or it is not a problem) here is that it is an imitation of HKT. Also we can add to `URItoKind` anything we want. Also we use just type aliases. So, yes I can write just `Interpreter<String>` but I wanted to make it look similar to Haskell.

Also I don't really know what different interpreter to write as we don't use `<A>` in `ToS` (but we also don't use it in Haskell implementation)

So, if you have any ideas I will change it.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  }
+}
+
+const arithInterpreter: ArithExpr<"Interpreter"> = {
```
> ArithExpr<"Interpreter">

What. How does it work? Please explain why are we putting a name of a field here.

Heimdell replied:
I guess, URI is an interface to string-like objects, but what is it here, exactly?

sashasashasasha151 replied:
Hm, I thought that HKT part explain it (do I need to explain in here again?) \
`URIS` is just keys(properties) of `URItoKind` where `URItoKind` is like map from keys to types \
Here we say that `ArithExpr` take `Expr extends URIS` and types in `ArithExpr` will be `Kind<Expr, number>` for example where `Kind` is a type from `URItoKind` with concrete keys(properties) from `URIS` \
:ghc: \
Let's think how to make it more understandable if `HTK` part doesn't describe it

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  toString: string;
+};
+
+declare module "fp-ts/lib/HKT" {
```
I just found I know anything about TS module system.

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,726 @@
+
+testExpr(arithInterpreter).interpret; // false
+testExpr(arithToS).toString; // (((23 + 12) > 170) && (35 > 47))
+```
````
The code above looks deceptively similar to haskell implementation, but in reality, there's a lot of magic in need of explanation.

Heimdell replied:
For the blocks of code, I usually consider them subjectively understandable, if I can sit and make an analogous code myself, given the example and an explanation.

Here, I can't, because there is no textual explanation.

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,726 @@
+  and: (l: Kind<Expr, boolean>, r: Kind<Expr, boolean>) => Kind<Expr, boolean>;
+};
+
+type Interpreter<A> = {
```
After reading the code block to the end, I start wondering what the `A` here really is. Plz explain how it is connected to other spells.

Heimdell reviewed (changes requested):
Please provide explanation (hand-wavy at least!) on how all of that black magic works.

Without it, this will scare the haskellers.

pva701 commented:
@sashasashasasha151 don't forget to acknowledge me, who gave your a hint about HKT/typelevel stuff in TS :putya:

sashasashasasha151 committed:
fixup! [PR-1190] Article about typescript for haskellers

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,916 @@
+
+Unlike Haskell in TypeScript types are not the center part of the language.
+They just help programmers to write stricter and more understandable code, making their life easier.
+You can use types like `any` and nobody pushes you into strict frames.
```
Do you mean anything specific by "strict frames"? I don't get it.

sashasashasasha151 replied:
Looks like I need word `framework` here

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,916 @@
+```
+
+Anyway, this is not an ideal solution for newtypes.
+There is a [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in issue with a better solution.
````
If that one is better, I have a question as a reader as to why I was provided with a worse one?

sashasashasasha151 replied:
It is better for like complex system and it is also complex for this example.

NaeosPsy replied:
`While this is the simplest solution for newtypes, it is not the best for complex systems. For a better (but more complex) one, check this [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in TypeScript's issues. `

NaeosPsy replied:
I think this would sound better, then.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,916 @@
+
+Unit types are subtypes of primitive types that contain exactly one primitive value.
+You can use it like enums setting unique string values.
+Or even use different values types.
```
This sentence is unclear, do you mean: "types with different values"?

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,916 @@
+type Result = true | "error" | 5;
+```
+
+You can pass it to the function and match by a concret type.
````
```suggestion
You can pass it to the function and match by the concrete type.
```

I assume

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,916 @@
+#### Unions
+
+Moving forward the next feature is Unions.
+Unions are not tagged, so it is just a condition of possible types.
```
condition is not really the best word I think, "they are just sets of possible types"

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,916 @@
+type Result = string | number | (() => string);
+```
+
+Here and in the example with Unit types if you already matched by some type compiler will know that it will not be present in the future.
````
word "matched" implies that it was matched successfully, I would use "failed to match" in this paragraph.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,916 @@
+
+#### Discriminated Unions or Datatype analogue
+
+Going deeper using Unions, it is possible to construct data type analog.
```
By going deeper with unions, it is possible to construct an analogue of Haskell data types?

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,916 @@
+};
+```
+
+#### Discriminated Unions or Datatype analogue
````
I don't think you need to write Union with capital U, please make it lowercase everywhere that I missed. :D

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,916 @@
+This is the very simple examples of working with ADT like structures in TypeScript.
+Using type parameters you are also able to make some more complex things like [such](https://github.com/pfgray/ts-adt) library for ADTs.
+But one of the most useful features of Haskell connected with ADTs is pattern matching.
+In allows you to disclose you structure comparing it with some pattern and based on this condition execute some code.
```
I was not able to understand this sentence, sorry. :D

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,916 @@
+
+TypeScript has two base primitives to work with immutability: const and readonly.
+
+The first one is used to prevent variable reference change.
```
Can you just say "to prevent reassigning variables"?

Heimdell replied:
Reference to the variable didn't change here, it is "on the same address", in terms of js references. The value did. Well, the reference can change, too, if you assign `a = b`, but the example shows us different scenario.

So, in the `const a` the `a` is _directly_ immutable (but you still can assign fields, I guess) and in `readonly a:` the field is inassignable.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,916 @@
+
+Also be careful passing object with readonly fields to function as it may be changed inside the function.
+It is happening because of information loss.
+Object with the type which include `readonly` is a subtype of the type without it and function argument with more common type know nothing about its subtype.
```
Do function arguments know things?

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,916 @@
+Let's introduce some libraries which may help you deal with this issue.
+The first one and the most powerful is [immer](https://immerjs.github.io/immer/).
+It is used for immutable state flow.
+Immutable data is not copied but shared in memory.
```
Is this a statement about all immutable data or about what immer lets you to do?

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,916 @@
+f("5")(5) // true;
+```
+
+As you may see in the last example, we can call function `f` and return function from it, applying in for the new argument.
````
Don't get the last clause of the sentence.

NaeosPsy committed:
Proof-reading until polymorphism

Grammar edits that don't require assistance.

sashasashasasha151 committed:
fixup! [PR-1190] Article about typescript for haskellers

sashasashasasha151 commented:
Looks like I added more code descriptions and answered/fixed all comments. It is possible that I missed smth or some parts need more detailed explanations

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1002 @@
+```
+
+Now let's implement `lookup` function which for key value pairs array and specific key will return the value associated with this key or `undefined`.
+We can write it polymorph passing comparator class which implements `Eq` for the key type.
````
I don't grasp this sentence. 

"We can make it polymorphic by passing a comparator class that implements `Eq` for the key type"?

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1002 @@
+Square brackets indicates computed names.
+And `in` operator goes over all field keys.
+
+`Partial` type adds `?` modifier for each property.
```
It seems that you now have two examples of the same thing. 

I would just do this section like this: 

* Syntax looks like this. 
* Example 
* Explanation of the syntax
* You can also remove properties
* Example

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1002 @@
+```
+
+Now, using mapped types we will create a type which will pick only some properties.
+Word `extends` here say that generic `K` will contain a subset of union properties keys `T`.
````
contain a subset of union properties keys `T` 

Something is not correct here grammar-wise, I can't parse it.  

Will contain a subset of the properties of `T`. <- does this work? 

In particular, is there a specific reason why you need to mention union all the time?

sashasashasasha151 replied:
> Will contain a subset of the properties of T.

Yes, it should work

> In particular, is there a specific reason why you need to mention union all the time?

So, mentioning `union` is like true name of what we have. I can mention it for example two times and then use just `set of properties` for example.

NaeosPsy replied:
Oh haha. :) Yeah, that should work.

NaeosPsy replied:
Grokked it and changed it.

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1002 @@
+So, for `T` which extends string we can produce getter specific name.
+
+```typescript
+type Getter<T extends string> = `get${Capitalize<T>}`;
````
Do you need to capitalize? It doesn't seem relevant for illustrating any getter, and then you won't need to explain it.

sashasashasasha151 replied:
>Do you need to capitalize?

Without capitalize it will look like `getsomething`. And usually you need to somehow separate the words.\
I really like this example as it show how you can change types using `` and how complex mapped types can be

NaeosPsy replied:
Ok, sure. 👍

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1002 @@
+
+To implement it lets firstly define helper type `Getter`.
+Capitalize is a build in type which may be applied for the string type.
+So, for `T` which extends string we can produce getter specific name.
```
Didn't get this sentence.

NaeosPsy replied:
@sashasashasasha151 it's still there, in particular, I don't understand what is a getter-specific name.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1002 @@
+Conditional types with union types are called Distributive Conditional Types and their work may look a little confused.
+But we will explain them on examples.
+
+Here is the base example of conditional types.
```
This should be right after the definition.

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1002 @@
+```
+
+Let's move to the Distributive Conditional Types.
+`Exclude` type may looks strange, but what if we will pass union types here?
````
I have to say I cannot understand this.

sashasashasasha151 replied:
Hm, I just say that `Exclude` is strange but with `union types` as type argument it is not strange and below is an explanation

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1002 @@
+
+#### Conditional types
+
+Another great feature is conditional types.
```
Structure is a bit off. 

I would just:

give a definition
show basic example 
tell that it can be used in conjunction with mapped types for that and that 
give a definition of distributed conditional types
show an example 

so that you are not jumping around stuff

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1002 @@
+You may see that TypeScript type system is not so bad as one may think.
+But it has one important limitation -- absence of kinds.
+Higher-Kinded Types allow us to write types which have their own type constructors as parameters.
+So, using generics instead of abstracting only over types we are able to abstract over types which already abstract over types.
```
:advanced-types: 

I think this can be written simpler.

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1002 @@
+};
+```
+
+Fortunately we are able to simulate kinds using defunctionalization which allows us to translate higher-order programs into a first-order language.
````
Is there some link to defunctionalization available? https://blog.sigplan.org/2019/12/30/defunctionalization-everybody-does-it-nobody-talks-about-it/ I found smth like this

sashasashasasha151 replied:
Hm, on wiki https://en.wikipedia.org/wiki/Defunctionalization

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1002 @@
+```
+
+So, now we are ready to present our kinds.
+They take an identifier property as the first type parameter and its type parameters for this identifier.
````
```suggestion
They take an identifier property as the first type parameter and its type parameters for this identifier as the second one. 
```

You have an unfinished sentence, I assume this is the full one.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1002 @@
+
+### Peano numbers
+
+Moving deeper we can think about type level programming and more concretely about computational on types.
```
computations on types?

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1002 @@
+
+This was a simple example what we can do with type level programming.
+Using such primitives we can implement different operations on types and even Fibonacci [implementation](https://mjj.io/2021/03/29/type-level-programming-in-typescript/) for example.
+Follow Mathias Jean Johansen [blog post](https://mjj.io/2021/03/29/type-level-programming-in-typescript/) or mentioned above Yuriy Bogomolov [examples](https://github.com/YBogomolov/talk-typelevel-ts/blob/master/src/typelevel.d.ts).
```
Linking one blog post two lines in a row is a big no-no, can you squeeze these two lines into one?

NaeosPsy committed:
Proof-read until GADTs

Silent grammar edits

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1002 @@
+Additionally, customers will be locked in when they will need to maintain such applications.
+On the other hand, TypeScript, which has a huge community and modern technologies out from the box, is a great decision.
+
+## Base concepts
```
```suggestion
## Basic concepts
```

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1005 @@
+
+Generalized algebraic data types in Haskell give us an ability to manually write types of the constructors.
+Having data type `D a` with type value `a` we are able to create constructors like `C :: Int -> D Int`.
+The key feature here is that Haskell may say us about equality of types `a` and `Int` here.
```
This line is a bit oddly worded, could you try to rephrase. In particular, "Haskell may say us".

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1005 @@
+The other great feature of this approach is that we can make nice eDSL using GADT.
+And again TypeScript will not give you such a good eDSL as a result, but it doesn't mean that we can't do it.
+
+TypeScript on the other hand doesn't have equality inference.
```
Missing the first hand, perhaps some changes messed the flow up.

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1005 @@
+
+We will use `fp-ts` here and later as it has a lot of build in useful types and functions.
+
+```typescript
````
What's happening here? No text points to it.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1005 @@
+To provide equality as a value we use Leibniz equality.
+Here is a simplified version of it.
+Interface `Equality` is a Hybrid Type which gives an ability for object act as a function.
+Then in `ArithExpr` we simplify `Equality<A, B>` into an identity `(a: A) => A`.
```
The explanation is too far from the thing explained.

sashasashasasha151 replied:
Hm, how to write it here? To somehow explain `Equality` I need to say about things described below but describing such things without `Equality` explanation is not really good

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1005 @@
+#### Tagless final eDSL
+
+Another way to implement eDSL is tagless final.
+Moving from data type to type class we are able to implement the same logic.
```
Don't get this line.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1005 @@
+Moving from data type to type class we are able to implement the same logic.
+
+Let's start from defining our type class.
+To make this we need already known `URIS` which is a union of all 1-arity defined in `URItoKind`.
```
Should it be 1-arities? (e.g. plural)

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1005 @@
+
+Now, we will present our types for which we will create `ArithExpr` instances.
+The first one will be just interpreter.
+And the second one will create string representation of expression.
```
You can insert the types here, I think. 

We need the description to be as maximally close to the code as possible.

NaeosPsy committed:
Finishing the first grammar read

Finished reading it the first time, will read the second time from start to finish later.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1004 @@
+Because we like Haskell, maybe we can write our Web application in a functional language (or even in Haskell)? The answer is yes.
+There is strongly-typed PureScript with Haskell-like syntax, purely functional Elm, Reflex framework for Haskell with GHCJS, F# with Fabel.
+
+Unfortunately, these don't give us proper flexibility in working with JS libraries (whose numbers are enormous) and take more time on fixing related problems.
```
What do you mean by "related problems"?

sashasashasasha151 replied:
Related to libraries management \
Choosing what libs are available, somehow import unavailable libs

NaeosPsy commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1004 @@
+};
+```
+
+Now, we will present our types for which we will create `ArithExpr` instances.
````
What does present mean here?

sashasashasasha151 replied:
Like show

NaeosPsy committed:
Grammarly + wording fixes

Second read of the article.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1015 @@
+There is a [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in issue with a better solution.
+And there is also a [library](https://github.com/sveseme/taghiro), which provides a prettier implementation for them.
+
+### Unit types, unions, data types, and pattern matching
```
Could this section be called ADTs in TypeScript?

sashasashasasha151 replied:
Not sure if it will be a right definition but in article title context I think we can do it

sashasashasasha151 replied:
Maybe just ADTs or Algebraic data types? To not explicitly mention TypeScript

NaeosPsy replied:
Yeah, that is fine.

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1015 @@
+
+Let's move to distributive conditional types.
+`Exclude` type may look strange, but what if we will pass union types here?
+It will simply exclude such types in union `T` which exists in union `U` setting `never` in other cases, where the [`never`](https://www.tutorialsteacher.com/typescript/typescript-never) type describes the type of values that will never occur.
```
What does "return never in other cases" mean here?

sashasashasasha151 replied:
Oh, not in other cases) \
Basically it is not returning smth but, 🤔, setting the type \
But `_ ? _ : _` construction is known as a ternary operator and saying `return`, as for me, sounds more understandable here \
So, I will change it, but maybe we can leave `return` here

NaeosPsy committed:
Edits after previewing on the website

sashasashasasha151 committed:
fixup! [PR-1190] Article about typescript for haskellers

NaeosPsy committed:
Edits to resolve some of my comments

NaeosPsy committed:
Resolving the ADT comment, small changes in the section

NaeosPsy committed:
Resolving some of the comments regarding the kind section

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1019 @@
+type Partial<T> = { [P in keyof T]?: T[P] };
+
+type A = { x: number; y: number };
+type PartialA = Partia<A>;
```
You sure this compiled? :) 

```suggestion
type PartialA = Partial<A>;
```

NaeosPsy commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1019 @@
+    [K in keyof T as Getter<Extract<K, string>>]: () => T[K];
+}
+
+type PointGetters = Getters<Point> // type PointValues = { getX: () => number; getY: () => number; }
```
I assume it should be like this: 

```suggestion
type PointGetters = Getters<Point> // type PointGetters = { getX: () => number; getY: () => number; }
```

NaeosPsy committed:
Editing Immutability, Mapped, conditional types

NaeosPsy committed:
Finish editing

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1038 @@
+
+const a: A = { x: 1, y: 2 };
+a.x = 2; // Index signature in type 'A' only permits reading.
+a.y = 1; // Index signature in type 'A' only permits reading.
```
Is it a clarifying comment, or some error that will be thrown here? If this is an error, is it a compile-time or runtime error? I would tell this clearly in places like this.

sashasashasasha151 replied:
There are no runtime errors in the article. I will mark it as error, but don't know if marking it as a compile time error will be useful

sashasashasasha151 replied:
And comments will be just comments, is it okay?

Martoon-00 commented:
One suggestion: I think it would be super beneficial if you provide a table that enlists the Haskell syntax and the corresponding TS syntax. 

Reading this article once is mostly enough to grasp the concepts, but syntax - people tend to forget it, and if you have a separate concise table that would be a nice resource that people will be able to consult constantly.

Something like

| Haskell      | TypeScript | Comment |
| ----------- | ----------- | ---------- |
| `Double`      | `number`       |  |
| `String` | `string` | |
| `type A = String` | ... | |
| `newtype A = A String` | ... |  |
| `<product type>` | ... | |
| `<sum type>`   | ... |        |
| `data A a = ...` | ... | |
...

Indeed, not every Haskell thing has a corresponding TS thing, in such cases you could mention the possible alternatives or that there is no corresponding functionality at all.

Martoon-00 commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1038 @@
+let twoElems: Vec<number, Succ<Succ<Zero>>> = pushVec(2, oneElem); // Ok
+let twoElemsInvalid: Vec<number, Succ<Zero>> = pushVec(2, oneElem); // error
+oneElem = twoElems; // error
+```
````
Just a question to satisfy my curiosity: could I use plain type-level number instead of Peano numbers? Or arithmetic operations and conditions do not work on plain type-level numbers?

sashasashasasha151 replied:
Hm, I was trying to understand how to do it, but I don't really understand how to write it. If you mean smth like:

```typescript
type Vec<A, N extends number> = N extends 0 ? Nil : Cons<A, N - 1>;
```

Then no, it is not possible

If you meant smth different may you please write an example on pseudo code

Martoon-00 replied:
Yes, exactly so. I see

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1038 @@
+Here the type class will describe possible actions, and by using instances, we will be able to write interpreters.
+
+Let's start by defining our type class.
+To do this, we need the already-mentioned `URIS`(which is a union of all 1-arity defined in `URItoKind`) and the already-mentioned `Kind`.
```
You can also say here, that you represent the type class as a dictionary of related methods carried in a mere type. That jump from Haskell type classes to the TS code with a record type was just too big.

sashasashasasha151 replied:
It was described in `HKT` part. But I think your explanation is better, so I can move it to the `HKT` part

sashasashasasha151 replied:
Ah, okay, I misunderstood you \

Martoon-00 replied:
Ah no, looks good to me now. Here you said "defining our type class" but TS has no typeclasses. And now when you in HKT section describe what you mean under TS typeclasses, "defining our type class" makes sense.

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1038 @@
+
+const collectionArray: Collection<Array> = {
+  create: () => [],
+  insert: (v) => (c: number[]) => [...c, v]
```
This `...` seems like a new, yet unexplained, syntax.

Martoon-00 reviewed (commented):
I reviewed the remaining part (starting from the mapped types) ~superficially, looks quite good overall. 

Things look really well explained (but not all of them yet, hope that resolving the remaining @Heimdell's comments will improve the situation). I mostly had no questions like "WTF is going on here". Many examples looked sort of familiar, so it was relatively easy to understand them (though will I be able to re-implement something here with the same ease - that would be an interesting question).

But for future reference, I again express my concern regarding the number of covered topics. The article starts from the basic things like syntax and types, then without giving a break it introduces tricky types where some decent understanding of basics is already required. Then the post picks up the pace even more, using concepts that are not explained anywhere, and keeping this pace is now hard for you as a writer.

NaeosPsy commented:
> One suggestion: I think it would be super beneficial if you provide a table that enlists the Haskell syntax and the corresponding TS syntax.


https://haskellfortypescriptdevs.fission.codes/part-i -- Just saw this, perhaps it can serve as partial inspiration.

sashasashasasha151 commented:
@Martoon-00

> One suggestion: I think it would be super beneficial if you provide a table that enlists the Haskell syntax and the corresponding TS syntax.

You mean in the end of the article (to summary part) or how? First of all when I started to write this article I thought that it will be for people who already knew the TypeScrips syntax. Then in review there was a good point to describe some non-trivial things and I added them.

But, hm, I don't really know what to add in this table if it will be in the beginning as the whole article is about translating Haskell code to TS code. I can do it in code parts like creating Haskell examples to all  TS examples (but not really know how do it)

Anyway I am not against your suggestion, just want to understand how to make it more useful and not just copy paste of the article

> But for future reference, I again express my concern regarding the number of covered topics.

I also thought about it. But in my head it looked like the first part is for all, there are no really hard concepts and the second part is for inspiration. I specially wrote that it is just examples with descriptions and for more info you need to go here and here. There is no reason to write a full article about them as only few amount of people will need it and even fewer amount of people will use it. \
So, the idea (as I introduced it to myself) was to write some kind of dictionary where you can quickly find information but not a fully explanation. For example when you google about parser combinators in Haskell you will not find good explanation, but with last Heitor blog you can really understand all that you need. And when you will google TS Mapped types you will find ton of articles about them. \
Not sure that you expected such answer. Maybe @NaeosPsy will add smth.

Martoon-00 commented:
> You mean in the end of the article (to summary part) or how?

Yes, I would put it at the end, or maybe even as a separate gist. This table should answer the question: "I want to declare X thing (e.g. polymorphic datatype) in TypeScript, remind me in 5 seconds which code should I write".

On the other hand, the resource that @NaeosPsy mentioned looks quite good, maybe leaving a link in some `Useful resources` section at the end of the post would be exactly sufficient.

> So, the idea (as I introduced it to myself) was to write some kind of dictionary where you can quickly find information but not a fully explanation.

I see, told this way it really makes sense, yeah.

NaeosPsy commented:
https://serokell.io/blog/parser-combinators-in-haskell

You can also take a look at how Heitor handled it in appendix of parser combinator article.

sashasashasasha151 committed:
fixup! [PR-1190] Article about typescript for haskellers

sashasashasasha151 commented:
@Martoon-00 @NaeosPsy I spent full day trying to understand how to make the table useful and informative \
So, the first question is about the number of content in it. Do I need to write only about topics in the article in it, or maybe some more info. \
In article there are some TypeScript specific things, so at first I tried to write TypeScript to Haskell table but in this case it is a little bit strange as this article is for Haskellers. On the other hand if I will write Haskell to TypeScript I will not cover some things. \
In case of TypeScript to Haskell there also is an advantage that we can add some syntax specific things to it \

And after all table has really small amount of space to write and there is no ability to add code blocks \
So, I propose to not create a table but maybe small part with just brief comparisons. In this case I will be able to add code blocks, add some TS specific info (and maybe TS specific syntax info) \
What do you think about it?

NaeosPsy commented:
@sashasashasasha151 

You can definitely use tables + code blocks. (See Vlad's translations from Haskell to Core, we just did HTML tables: https://serokell.io/blog/haskell-to-core) 

A TypeScript to Haskell guide would be quite odd in this case. This particular item should serve as an aid for memory, so it also doesn't have to contain explanations of TypeScript stuff. 

Basically, you can look at it as creating a "cheatsheet" for **your** article. In case something is too long, you can just use an anchor link to the corresponding section.

You can also link to the other one at the bottom of it if you wish.


Edit: I've looked at the one you've already created, and it is okay, just that there is no space for the code. If you had only two columns at max, it would be solved. I think the table can be restructured or divided so that it works like that.

NaeosPsy commented:
```html

<table>

<tr>
	<th>Concept</th>	
	<th>Haskell</th>
	<th>TypeScript</th>
</tr>
	
<tr>
<td>

[Type alias](#type-aliases)	

</td>	
	
<td>	
		
```haskell
type Email = String 
``` 
</td>
<td>	
		
```typescript
type Email = string 
``` 
</td>
</tr>

</table>
```

If you do it like this, it should be fine, the table will extend if necessary and will be scrollable to left and right.

sashasashasasha151 committed:
fixup! [PR-1190] Article about typescript for haskellers

sashasashasasha151 commented:
@Martoon-00 @NaeosPsy I made a table, not sure if it will look good on the site. Also I don't know if it contain information you thought about.

NaeosPsy committed:
Fix parenthesis in one place

NaeosPsy committed:
Proof-read changes, fix PointValues/PointGetters

Martoon-00 commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1649 @@
+
+And it is how type classes will look in TypeScript.
+We represent them as a dictionary of related methods carried in a mere type.
+Our `Collection` is a type class which work with one arity constructors which we defined in `URItoKind`.
```
Formally it is not a type class, it is a type. I would say 

```suggestion
Our `Collection` is sort of a type class which work with one arity constructors which we defined in `URItoKind`.
```

or

```suggestion
Our `Collection` is effectively a type class which work with one arity constructors which we defined in `URItoKind`.
```

Martoon-00 commented:
The tables are just awesome! Thanks for adding them :+1:

---

One more thing I'd like to see there: does TS have something like `TypeError`? So that I could write this

```ts
type G<A> = A extends number ? boolean : A extends string ? boolean : blowUp;
```

similar to

```hs
type family G a where
  G Int = Bool
  B String = Bool
  -- no "otherwise" clause
```
?

sashasashasasha151 commented:
@Martoon-00
`type G<A> = A extends number ? boolean : A extends string ? boolean : never;`?

serokellcao commented:
Gints, are we gonna follow up on our collab with Brooklyn with this?

On Fri, 11 Feb 2022, 01:21 Sasha Pakulev, ***@***.***> wrote:

> @Martoon-00 <https://github.com/Martoon-00>
> type G<A> = A extends number ? boolean : A extends string ? boolean :
> never;?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/serokell/blog-posts/pull/5#issuecomment-1035700187>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK2DJQZ755RZ7PZ5PGEXBIDU2RQBZANCNFSM5JCB5FMA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

sashasashasasha151 committed:
fixup! [PR-1190] Article about typescript for haskellers

Martoon-00 commented:
> @Martoon-00 `type G<A> = A extends number ? boolean : A extends string ? boolean : never;`?

Ah no, this would correspond to `_ -> Void`, which maps to an uninhabited but valid type, and I want to forbid all other cases for `G` so that `G Float` is a type error.

sashasashasasha151 commented:
`type G<A extends number | string> = A extends number ? boolean : boolean;`

Martoon-00 commented:
> `type G<A extends number | string> = A extends number ? boolean : boolean;`

Ooh right, and this perhaps looks even better than for Haskell.

Could you include this? Like, here the correspondence to Haskell is not 1-to-1, would be nice to have this mentioned as a hint.

sashasashasasha151 committed:
fixup! [PR-1190] Article about typescript for haskellers

Martoon-00 reviewed (approved):
Thanks for addressing my comments!

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1655 @@
+
+In the first `if`, we pass `typeof result === "number"`, so we know that it may be only `function` or `string` in the code below.
+And we can run `length.toString()` on this value (even if `number` doesn't have such a property) since both `function` and `string` have the property `length`.
+But we can't `call` this value because `string` is not callable. 
```
But can we call it, if all constituents of the union are callable? Will it infer the result to be union of function result types?

sashasashasasha151 replied:
Yes, added note

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1655 @@
+
+```typescript
+const resultInterpreter = (result: Result): string | undefined => {
+  if (typeof result === "number") {
````
So, `if` statement is enriched with case analysis, when is used with `typeof result === "number"`.

Will `typeof result === "number" || typeof result === "string"` do the trick, leaving type to be `function` in `else`-block?

sashasashasasha151 replied:
Yes, added note

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1655 @@
+return match(result)
+  .with({ type: 'error' }, (res) => `<p>Oups! An error occured</p>`)
+  .with({ type: 'ok', data: { type: 'text' } }, (res) => `<p>${res.data.content}</p>`)
+  .with({ type: 'ok', data: { type: 'img', src: select() } }, (src) => `<img src=${src} />`)
```
Is the rule "`res` is the matched payload _or_ arguments are the ones chosen with `select()`"?

sashasashasasha151 replied:
Looks like yes, but `ts-pattern` have a lot of things, it is a simple example just for design understanding

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1655 @@
+a.x; // 2
+```
+
+For this reason, you should always write proper types of arguments.
````
What do you mean?

sashasashasasha151 replied:
I think it was moved here by mistake

Heimdell commented on typescript-for-haskellers/article.md:
````diff
@@ -0,0 +1,1655 @@
+});
+```
+
+In the code example above, `produce` is a base primitive for working with `immer`.
````
Does it remove all `readonly`-limitations from type of `a` via `Writable<>`?

sashasashasasha151 replied:
I don't really know how it works under the hood, but what for it may need `Writable`? Search in their repo by `Writable` said that there is no `Writable`

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1655 @@
+
+#### Parametric polymorphism
+
+Parametric polymorphism allows us to write abstract functions or data types that don't depend on their type.
```
```suggestion
Parametric polymorphism allows us to write abstract functions that, for instance, don't depend on concrete types of particular arguments, and types that can abstract out parts of itself as type parameters.
```

Heimdell commented on typescript-for-haskellers/article.md:
```diff
@@ -0,0 +1,1655 @@
+  and: (l: ToS<boolean>, r: ToS<boolean>) => {
+    return { toString: `(${l.toString} && ${r.toString})` };
+  },
+};
```
My wild guess is that `"ToS"` here is a marker or has the same role as
```
instance IsExpr ToS where
```
in haskell.

sashasashasasha151 replied:
If I understand you right the answer is yes \
I don't really like how it looks, but I wanted to make it looks like in haskell and didn't want to invent more complex examples as it is already complex enough. So, I thought a lot about it and as it is kind of advanced part I think we can leave it like this

Heimdell reviewed (approved):
Looks overall good.

Some demon-conjuration-spells are not clearly explained, but most of them are. I think readers should start poking into TS just out of initial contact shock.

The appendix with haskell-TS correspondence is nice.

sashasashasasha151 committed:
fixup! [PR-1190] Article about typescript for haskellers

NaeosPsy committed:
Sync with website
